libbeat/processors/annotate/kubernetes/indexing.go

```diff
@@ -0,0 +1,256 @@
+		}
+	}
+	return indexes
+}
+
+// MetadataIndex returns the index string for the first matcher from the Registry returning one
+func (r *Register) MetadataIndex(event common.MapStr) string {
+	r.RLock()
+	defer r.RUnlock()
+	for _, matcher := range r.matchers {
+		index := matcher.MetadataIndex(event)
+		if index != "" {
+			return index
+		}
+	}
+
+	// No index returned
+	return ""
+}
+
+// Implement default indexer:
+
+// GenMetadata generates default metadata for the given pod
+func GenMetadata(pod *corev1.Pod) common.MapStr {
+	labelMap := common.MapStr{}
+	for k, v := range pod.Metadata.Labels {
+		labelMap[k] = v
+	}
+	return common.MapStr{
+		"pod":       pod.Metadata.GetName(),
+		"namespace": pod.Metadata.GetNamespace(),
+		"labels":    labelMap,
+	}
+}
+
+// PodNameIndexer implements default indexer based on pod name
+type PodNameIndexer struct{}
+
+func (p *PodNameIndexer) GetMetadata(pod *corev1.Pod) []MetadataIndex {
+	data := GenMetadata(pod)
+	return []MetadataIndex{
+		{
+			Index: pod.Metadata.GetName(),
+			Data:  data,
+		},
+	}
+}
+
+func (p *PodNameIndexer) GetIndexes(pod *corev1.Pod) []string {
+	return []string{pod.Metadata.GetName()}
+}
+
+// ContainerIndexer indexes pods based on all their containers IDs
+type ContainerIndexer struct{}
+
+func (c *ContainerIndexer) GetMetadata(pod *corev1.Pod) []MetadataIndex {
+	commonMeta := GenMetadata(pod)
+	containers := c.GetIndexes(pod)
+	var metadata []MetadataIndex
+	for i := 0; i < len(containers); i++ {
+		containerMeta := commonMeta.Clone()
+		containerMeta["container"] = pod.Status.ContainerStatuses[i].Name
+		metadata = append(metadata, MetadataIndex{
+			Index: containers[i],
+			Data:  containerMeta,
+		})
+	}
+
+	return metadata
+}
+
+func (c *ContainerIndexer) GetIndexes(pod *corev1.Pod) []string {
+	var containers []string
+	for _, status := range pod.Status.ContainerStatuses {
+		cID := status.GetContainerID()
+		if cID != "" {
+			parts := strings.Split(cID, "//")
+			if len(parts) == 2 {
+				containers = append(containers, parts[1])
+			}
+		}
+	}
+	return containers
+}
+
+type FieldMatcher struct {
+	MatchFields []string
+}
+
+func newFieldMatcher(cfg common.Config) Matcher {
+	config := struct {
+		LookupFields []string `config:"lookup_fields"`
+	}{}
+
+	err := cfg.Unpack(&config)
+	if err != nil || len(config.LookupFields) == 0 {
+		logp.Err("fail to unpack the `lookup_fields` configuration: %s", err)
+		return nil
+	}
+
+	return &FieldMatcher{MatchFields: config.LookupFields}
+}
+
+func (f *FieldMatcher) MetadataIndex(event common.MapStr) string {
+
+	for _, field := range f.MatchFields {
+		keyIface, err := event.GetValue(field)
+		if err != nil {
+			key, ok := keyIface.(string)
+			if ok {
+				return key
+			}
+		}
+	}
+
+	return ""
+}
+
+//TODO: Move this into filebeat
+type LogPathMatcher struct {
+	LogsPath string
+}
+
+func newLogsPathMatcher(cfg common.Config) Matcher {
+	config := struct {
+		LogsPath string `config:"logs_path"`
```

Comment:
Assuming this is in filebeat, that would mean the filebeat processors have an additional config option?